File: src/ui/web/codemirror_editor.py
"""CodeMirror 6 editor component for NiceGUI.

This module provides a custom NiceGUI component wrapping CodeMirror 6,
enabling rich text editing with decorations, markers, and syntax highlighting.

Features:
- Find highlighting (yellow background)
- Breakpoint markers (red gutter markers)
- Current statement highlighting (green/blue background for step debugging)
- Line numbers
- Proper text editing with undo/redo
"""

from nicegui import ui
from typing import Callable, Optional


class CodeMirrorEditor(ui.element, component='codemirror_editor.js'):
    """CodeMirror 6 editor component.

    This component provides a rich code editor with support for:
    - Text decorations (find highlighting, current statement)
    - Line gutter markers (breakpoints)
    - Programmatic scroll control
    - Event callbacks

    Example:
        editor = CodeMirrorEditor(
            value="10 PRINT 'Hello'\n20 END",
            on_change=lambda e: print(f"Content: {e.sender.value}")
        )

        # Highlight find results
        editor.add_find_highlight(line=0, start_col=3, end_col=8)

        # Add breakpoint marker
        editor.add_breakpoint(line_num=10)

        # Highlight current executing statement
        editor.set_current_statement(line_num=10)
    """

    def __init__(self,
                 value: str = '',
                 on_change: Optional[Callable] = None,
                 readonly: bool = False) -> None:
        """Initialize CodeMirror editor.

        Args:
            value: Initial text content
            on_change: Callback when text changes (receives event with .sender.value)
            readonly: Whether editor is read-only
        """
        super().__init__()
        self._value = value
        self._readonly = readonly
        self._props['value'] = value
        self._props['readonly'] = readonly

        # Internal handler to keep _value in sync with user edits
        def _internal_change_handler(e):
            self._value = e.args  # CodeMirror sends new value as args
            if on_change:
                on_change(e)

        if on_change:
            self.on('change', _internal_change_handler)
        else:
            self.on('change', lambda e: setattr(self, '_value', e.args))

    @property
    def value(self) -> str:
        """Get current editor content."""
        return self._value

    @value.setter
    def value(self, text: str) -> None:
        """Set editor content."""
        self._value = text
        self._props['value'] = text
        self.run_method('setValue', text)

    def get_value(self) -> str:
        """Get current editor content (async)."""
        return self.run_method('getValue')

    def set_value(self, text: str) -> None:
        """Set editor content."""
        self.value = text

    def add_find_highlight(self, line: int, start_col: int, end_col: int) -> None:
        """Add yellow highlight for find results.

        Args:
            line: 0-based line number
            start_col: Start column (0-based)
            end_col: End column (exclusive)
        """
        self.run_method('addFindHighlight', line, start_col, end_col)

    def clear_find_highlights(self) -> None:
        """Remove all find highlights."""
        self.run_method('clearFindHighlights')

    def add_breakpoint(self, line_num: int) -> None:
        """Add breakpoint marker to line gutter.

        Args:
            line_num: BASIC line number (e.g., 100, 200)
        """
        self.run_method('addBreakpoint', line_num)

    def remove_breakpoint(self, line_num: int) -> None:
        """Remove breakpoint marker from line gutter.

        Args:
            line_num: BASIC line number
        """
        self.run_method('removeBreakpoint', line_num)

    def clear_breakpoints(self) -> None:
        """Remove all breakpoint markers."""
        self.run_method('clearBreakpoints')

    def set_current_statement(self, line_num: Optional[int]) -> None:
        """Highlight current executing statement during step debugging.

        Args:
            line_num: BASIC line number, or None to clear highlight
        """
        self.run_method('setCurrentStatement', line_num)

    def scroll_to_line(self, line: int) -> None:
        """Scroll to and center the specified line.

        Args:
            line: 0-based line number
        """
        self.run_method('scrollToLine', line)

    def get_cursor_position(self) -> dict:
        """Get current cursor position.

        Returns:
            Dict with 'line' and 'column' (both 0-based)
        """
        return self.run_method('getCursorPosition')

    def set_readonly(self, readonly: bool) -> None:
        """Set editor read-only state.

        Args:
            readonly: True to make read-only, False to allow editing
        """
        self._readonly = readonly
        self._props['readonly'] = readonly
        self.run_method('setReadonly', readonly)

File: src/ui/web/codemirror_editor.js
/**
 * CodeMirror 6 Editor Component for NiceGUI
 *
 * Provides a rich code editor with:
 * - Find highlighting (yellow background)
 * - Breakpoint markers (red gutter markers)
 * - Current statement highlighting (green/blue background)
 * - Line numbers
 */

import { EditorView, keymap, lineNumbers, highlightActiveLine } from "@codemirror/view";
import { EditorState, StateEffect, StateField } from "@codemirror/state";
import { defaultKeymap, history, historyKeymap } from "@codemirror/commands";
import { Decoration, WidgetType } from "@codemirror/view";

export default {
    template: '<div></div>',

    mounted() {
        // Extension for find highlights
        const findHighlightEffect = StateEffect.define();
        const findHighlightField = StateField.define({
            create() { return Decoration.none; },
            update(decorations, tr) {
                decorations = decorations.map(tr.changes);
                for (let effect of tr.effects) {
                    if (effect.is(findHighlightEffect)) {
                        decorations = effect.value;
                    }
                }
                return decorations;
            },
            provide: f => EditorView.decorations.from(f)
        });

        // Extension for breakpoint markers
        const breakpointEffect = StateEffect.define();
        const breakpointField = StateField.define({
            create() { return Decoration.none; },
            update(decorations, tr) {
                decorations = decorations.map(tr.changes);
                for (let effect of tr.effects) {
                    if (effect.is(breakpointEffect)) {
                        decorations = effect.value;
                    }
                }
                return decorations;
            },
            provide: f => EditorView.decorations.from(f)
        });

        // Extension for current statement highlight
        const currentStatementEffect = StateEffect.define();
        const currentStatementField = StateField.define({
            create() { return Decoration.none; },
            update(decorations, tr) {
                decorations = decorations.map(tr.changes);
                for (let effect of tr.effects) {
                    if (effect.is(currentStatementEffect)) {
                        decorations = effect.value;
                    }
                }
                return decorations;
            },
            provide: f => EditorView.decorations.from(f)
        });

        // Create editor
        this.view = new EditorView({
            state: EditorState.create({
                doc: this.value || '',
                extensions: [
                    lineNumbers(),
                    history(),
                    highlightActiveLine(),
                    keymap.of([...defaultKeymap, ...historyKeymap]),
                    findHighlightField,
                    breakpointField,
                    currentStatementField,
                    EditorView.updateListener.of((update) => {
                        if (update.docChanged) {
                            this.value = update.state.doc.toString();
                            this.$emit('change', this.value);
                        }
                    }),
                    EditorState.readOnly.of(this.readonly || false)
                ]
            }),
            parent: this.$el
        });

        // Store effects for later use
        this.findHighlightEffect = findHighlightEffect;
        this.breakpointEffect = breakpointEffect;
        this.currentStatementEffect = currentStatementEffect;

        // Store current decorations
        this.findHighlights = [];
        this.breakpoints = new Set();
        this.currentStatement = null;
    },

    beforeUnmount() {
        this.view.destroy();
    },

    methods: {
        setValue(text) {
            this.view.dispatch({
                changes: { from: 0, to: this.view.state.doc.length, insert: text }
            });
            this.value = text;
        },

        getValue() {
            return this.view.state.doc.toString();
        },

        addFindHighlight(line, startCol, endCol) {
            const doc = this.view.state.doc;
            const lineObj = doc.line(line + 1);  // CodeMirror lines are 1-based
            const from = lineObj.from + startCol;
            const to = lineObj.from + endCol;

            this.findHighlights.push({ line, startCol, endCol });

            const decorations = this.findHighlights.map(h => {
                const lineObj = doc.line(h.line + 1);
                const from = lineObj.from + h.startCol;
                const to = lineObj.from + h.endCol;
                return Decoration.mark({ class: 'cm-find-highlight' }).range(from, to);
            });

            this.view.dispatch({
                effects: this.findHighlightEffect.of(Decoration.set(decorations))
            });
        },

        clearFindHighlights() {
            this.findHighlights = [];
            this.view.dispatch({
                effects: this.findHighlightEffect.of(Decoration.none)
            });
        },

        addBreakpoint(lineNum) {
            this.breakpoints.add(lineNum);
            this.updateBreakpoints();
        },

        removeBreakpoint(lineNum) {
            this.breakpoints.delete(lineNum);
            this.updateBreakpoints();
        },

        clearBreakpoints() {
            this.breakpoints.clear();
            this.updateBreakpoints();
        },

        updateBreakpoints() {
            const doc = this.view.state.doc;
            const decorations = [];

            // Find lines with BASIC line numbers matching breakpoints
            for (let i = 1; i <= doc.lines; i++) {
                const lineText = doc.line(i).text;
                const match = lineText.match(/^\s*(\d+)\s/);
                if (match) {
                    const basicLineNum = parseInt(match[1]);
                    if (this.breakpoints.has(basicLineNum)) {
                        // Add line decoration (red background)
                        decorations.push(
                            Decoration.line({ class: 'cm-breakpoint-line' })
                                .range(doc.line(i).from)
                        );
                    }
                }
            }

            this.view.dispatch({
                effects: this.breakpointEffect.of(Decoration.set(decorations))
            });
        },

        setCurrentStatement(lineNum) {
            this.currentStatement = lineNum;

            if (lineNum === null) {
                this.view.dispatch({
                    effects: this.currentStatementEffect.of(Decoration.none)
                });
                return;
            }

            const doc = this.view.state.doc;
            const decorations = [];

            // Find line with matching BASIC line number
            for (let i = 1; i <= doc.lines; i++) {
                const lineText = doc.line(i).text;
                const match = lineText.match(/^\s*(\d+)\s/);
                if (match && parseInt(match[1]) === lineNum) {
                    decorations.push(
                        Decoration.line({ class: 'cm-current-statement' })
                            .range(doc.line(i).from)
                    );
                    break;
                }
            }

            this.view.dispatch({
                effects: this.currentStatementEffect.of(Decoration.set(decorations))
            });
        },

        scrollToLine(line) {
            const doc = this.view.state.doc;
            const lineObj = doc.line(line + 1);  // CodeMirror lines are 1-based
            this.view.dispatch({
                effects: EditorView.scrollIntoView(lineObj.from, { y: 'center' })
            });
        },

        getCursorPosition() {
            const pos = this.view.state.selection.main.head;
            const line = this.view.state.doc.lineAt(pos);
            return {
                line: line.number - 1,  // Convert to 0-based
                column: pos - line.from
            };
        },

        setReadonly(readonly) {
            this.view.dispatch({
                effects: StateEffect.reconfigure.of([
                    EditorState.readOnly.of(readonly)
                ])
            });
        }
    },

    watch: {
        value(newValue) {
            if (this.view && newValue !== this.view.state.doc.toString()) {
                this.setValue(newValue);
            }
        },
        readonly(newValue) {
            if (this.view) {
                this.setReadonly(newValue);
            }
        }
    }
};
